### PR TITLE
[MLIR][Presburger] Implement IntegerRelation::mergeAndAlignSymbols

### PR DESCRIPTION
**Pull request by iambrj**

### CODE REVIEW

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 1]
```diff
@@ -1288,6 +1288,37 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
...
+    } else {
+      other.insertVar(VarKind::Symbol, i);
+      other.space.getId(VarKind::Symbol, i) = identifier;
+    }
+    ++i;
+  }
+
+  // Finally add identifiers that are in `other`, but not in `this` to `this`.
+  for (unsigned e = other.getNumVarKind(VarKind::Symbol); i < e; ++i) {
+    insertVar(VarKind::Symbol, i);
```
Comment (Groverkss) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1317]: We should add a insertVar variant with Identifier at some point. Not in this patch, but just wanted to point it out.

File: mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp [hunk 2]
```diff
@@ -207,3 +207,86 @@ TEST(IntegerRelationTest, swapVar) {
   EXPECT_TRUE(swappedSpace.getId(VarKind::Symbol, 1)
                   .isEqual(space.getId(VarKind::Domain, 1)));
 }
+
+TEST(IntegerRelationTest, mergeAndAlignSymbols) {
+  IntegerRelation rel =
+      parseRelationFromSet("(x, y, z, a, b, c)[N, Q] : (a - x - y == 0, "
+                           "x >= 0, N - b >= 0, y >= 0, Q - y >= 0)",
+                           3);
```
Comment (Groverkss) [inline mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp:215]: Can you add two tests:

- Unrelated identifiers: [A, B, C, D], [E, F, G, H]
- Unrelated prefix, then related identifiers: [A, B, C, D], [E, F, C, D]

**Review by Groverkss — CHANGES_REQUESTED** [review]

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 3]
```diff
@@ -1288,6 +1288,37 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
   removeVar(posB);
 }
 
+void IntegerRelation::mergeAndAlignSymbols(IntegerRelation &other) {
+  assert(space.isUsingIds() && other.space.isUsingIds() &&
+         "Both relations need to have identifers to merge & align");
+
+  // First merge & align identifiers into `other` from `this`.
```
Comment (Groverkss) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1295]: nit: & -> and

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 4]
```diff
@@ -1288,6 +1288,37 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
...
+  unsigned i = 0;
+  for (const Identifier identifier : space.getIds(VarKind::Symbol)) {
+    // If the identifier exists in `other`, then align it; otherwise insert it
+    // assuming it is a new identifier. Search in `other` starting at position
```
Comment (Groverkss) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1299]: The first sentance should be outside this for loop.

Comment (Groverkss) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1299]: Best place would be to put it as function implementation documentation (putting it on line 1290)

File: mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp [hunk 5]
```diff
@@ -207,3 +207,243 @@ TEST(IntegerRelationTest, swapVar) {
...
+            Identifier(&otherIdentifiers[1]));
+  EXPECT_EQ(otherSpace.getId(VarKind::Domain, 2),
+            Identifier(&otherIdentifiers[2]));
+  EXPECT_EQ(otherSpace.getId(VarKind::Range, 0),
+            Identifier(&otherIdentifiers[3]));
+  EXPECT_EQ(otherSpace.getId(VarKind::Range, 1),
+            Identifier(&otherIdentifiers[4]));
+}
+
+TEST(IntegerRelationTest, mergeAndAlignSymbols2) {
```
Comment (Groverkss) [inline mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp:294]: A better name explicitly stating what kind of test case it is checking would be nicer. Also a comment on what this does.

File: mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp [hunk 6]
```diff
@@ -207,3 +207,243 @@ TEST(IntegerRelationTest, swapVar) {
...
+            Identifier(&otherIdentifiers[0]));
+  EXPECT_EQ(otherSpace.getId(VarKind::Domain, 1),
+            Identifier(&otherIdentifiers[1]));
+  EXPECT_EQ(otherSpace.getId(VarKind::Range, 0),
+            Identifier(&otherIdentifiers[2]));
+  EXPECT_EQ(otherSpace.getId(VarKind::Range, 1),
+            Identifier(&otherIdentifiers[3]));
+}
+
+TEST(IntegerRelationTest, mergeAndAlignSymbols3) {
```
Comment (Groverkss) [inline mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp:375]: Same here.

File: mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp [hunk 7]
```diff
@@ -207,3 +207,243 @@ TEST(IntegerRelationTest, swapVar) {
...
+  otherSpace.getId(VarKind::Symbol, 3) = Identifier(&identifiers[6]);
+
+  rel.setSpace(space);
+  otherRel.setSpace(otherSpace);
+  rel.mergeAndAlignSymbols(otherRel);
+
+  space = rel.getSpace();
+  otherSpace = otherRel.getSpace();
+
+  // Check if merge & align is successful.
```
Comment (Groverkss) [inline mlir/unittests/Analysis/Presburger/IntegerRelationTest.cpp:415]: nit: & -> and

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 8]
```diff
@@ -1288,6 +1288,38 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
   removeVar(posB);
 }
 
+/// First merge and align identifiers into `other` from `this`. If an identifier
```
Comment (Groverkss) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1291]: "First" here describes the entire function right? Why is it "first" then.

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 9]
```diff
@@ -1288,6 +1288,38 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
...
+/// exists in `other`, then align it; otherwise insert it assuming it is a new
+/// identifier. Then add identifiers that are in `other`, but not in `this` to
+/// `this`.
```
Comment (Groverkss) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1294]: This is a little confusing... Can we write it steps?

**Review by Groverkss — CHANGES_REQUESTED** [review]
Just some documentation requests. Once that is done, we can land this.

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp (hunk 8 above)
Comment (iambrj) [reply mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1291]: Ah, I had written "first" as the implementation can be broken down into two parts, and "first" is the first parts:
1. Merge and align identifiers into `other` from `this`
2. Add identifiers that are in `other`, but not `this` to `this`

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 10]
```diff
@@ -1288,6 +1288,38 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
...
+void IntegerRelation::mergeAndAlignSymbols(IntegerRelation &other) {
+  assert(space.isUsingIds() && other.space.isUsingIds() &&
+         "Both relations need to have identifers to merge and align");
```
Comment (Abhinav271828) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1297]: nit: I believe [the LLVM standard](https://llvm.org/docs/CodingStandards.html#error-and-warning-messages) is for error messages to start with a lowercase letter.

File: mlir/lib/Analysis/Presburger/IntegerRelation.cpp [hunk 11]
```diff
@@ -1288,6 +1288,38 @@ void IntegerRelation::eliminateRedundantLocalVar(unsigned posA, unsigned posB) {
...
+
+  unsigned i = 0;
+  for (const Identifier identifier : space.getIds(VarKind::Symbol)) {
+    // Search in `other` starting at position `i` since the left of `i` is
+    // aligned.
+    auto *findBegin = other.space.getIds(VarKind::Symbol).begin() + i;
+    auto *findEnd = other.space.getIds(VarKind::Symbol).end();
+    auto *itr = std::find(findBegin, findEnd, identifier);
```
Comment (Abhinav271828) [inline mlir/lib/Analysis/Presburger/IntegerRelation.cpp:1305]: if the type name is not too large, it may help to write it explicitly?

**Review by Abhinav271828 — APPROVED** [review]

**Review by Groverkss — APPROVED** [review]
LGTM

**Pull request merged**